source/opt/const_folding_rules.cpp

```diff
@@ -341,6 +341,70 @@ ConstantFoldingRule FoldVectorTimesScalar() {
   };
 }
 
+// Returns to the constant that results from tranposing |matrix|. The result
+// will have type |result_type|, and |matrix| must exist in |context|. The
+// result constant will also exist in |context|.
+const analysis::Constant* TransposeMatrix(const analysis::Constant* matrix,
+                                          analysis::Matrix* result_type,
+                                          IRContext* context) {
+  analysis::ConstantManager* const_mgr = context->get_constant_mgr();
+  const analysis::NullConstant* matrix_as_null = matrix->AsNullConstant();
+  if (matrix_as_null != nullptr) {
+    matrix = const_mgr->GetNullCompositeConstant(matrix->type());
```

Comment:
If the matrix is a null composite constant, couldn't we do `return const_mgr->GetNullCompositeConstant(result_type)` ?

Reply:
Yes. This is what happens when you do things in multiple passes. My first pass was trying to avoid passing in the result type, so I did not use it here. I later realized I needed it at the end, and added it.